src/overlap.rs

```diff
@@ -0,0 +1,88 @@
+use std::sync::Arc;
+
+use itertools::Itertools;
+
+use errors::*;
+use ir::*;
+use solve::solver::Solver;
+
+impl Program {
+    pub fn check_overlapping_impls(&self) -> Result<()> {
+        let mut solver = Solver::new(&Arc::new(self.environment()), 10);
+
+        // Create a vector of references to impl datums, sorted by trait ref
+        let impl_data = self.impl_data.values().sorted_by(|lhs, rhs| {
+            lhs.binders.value.trait_ref.trait_id.cmp(&rhs.binders.value.trait_ref.trait_id)
+        });
+
+        // Group impls by trait.
+        let impl_groupings = impl_data.into_iter().group_by(|impl_datum| {
+            impl_datum.binders.value.trait_ref.trait_id
+        });
+
+        for (trait_id, impls) in &impl_groupings {
+            let impls: Vec<&ImplDatum> = impls.collect();
+
+            // For each pair, check their overlap by generating an "intersection"
+            // goal. In this case, success is an error - it means that there is at
+            // least one type in the intersection of these two impls.
+            for (lhs, rhs) in impls.into_iter().tuple_combinations() {
+                match solver.solve_goal(intersection_of(lhs, rhs)) {
+                    Ok(_)   => {
+                        let trait_id = self.type_kinds.get(&trait_id).unwrap().name;
+                        Err(Error::from_kind(ErrorKind::OverlappingImpls(trait_id)))
+                    }
+                    Err(_)  => Ok(())
+                }?;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+// The goal to test overlap.
+//
+// If this goal succeeds, these two impls overlap.
```

Comment:
Even though this function is very nicely commented, I think it'd be *that much more* helpful to give an example of two impls and the resulting goal from this function.